### ISSUE TITLE
Feature: Compatability with Quarkus.

### ISSUE THREAD

**Issue by whitingjr** [jruby/jruby#5665]
Hi,
 Is there any thoughts on making JRuby compatible with [Quarkus](https://developers.redhat.com/blog/2019/03/20/quarkus-0-12-0-released/) ?
 It makes jvm boot speed issues go away. Great for command line gems.

**Comment by headius** [comment]
Yes, we'd love to work with Quarkus! I actually have been hoping to try compiling a trimmed version of JRuby that uses less (or zero) reflection and no invokedynamic. JRuby 1.7 could also fully AOT compile Ruby code to JVM bytecode (9k does not do that yet) so it might be possible to precompile JRuby *and* your Ruby code.

Are you interested in helping us identify issues and getting JRuby working on Quarkus?

**Comment by whitingjr** [comment]
@headius thanks for the offer to help out. Beyond trying the basics and reporting any issues that's about all I can do. I've enough on my plate at the moment as it is.

**Comment by whitingjr** [comment]
@headius In an attempt to help progress the project forward I'll ask a question. About strategy. The question is a difficult one that involves a non-trivial amount of work. That I can't commit to solving.
 When I see the direction of OpenJDK and in particular [JSR 260](https://openjdk.java.net/jeps/260) becoming rigidly enforced it presents an issue for JRuby. The adoption of Quarkus hinges on no dependencies on reflection. The question I have (wearing my observer Hat) is what is the strategy for removing the reliance on Reflection in the JRuby project?

**Comment by headius** [comment]
The adoption of Quarkus depends on *predictable* reflection. It is always possible for us to mark up code or configure the build to know about the (relatively few) invasive reflections we have to do. More interesting will be to see how closely the class library followed OpenJDK...for AOT there *must* be some internal classes they had to rework.

My general thoughts for supporting Quarkus:

* Push forward plans to do ahead-of-time code generation, from precompiled Ruby code to pre-generated specialized objects to *direct* call sites that assume one or two static targets before doing a dynamic lookup.

This will allow us to have most call sites be direct, possibly inlined and specialized by the GraalVM AOT compiler, while avoiding most reflection.

* Rig up a system for users to specify the classes they reflect from Ruby code (or to gather that information during e.g. a test run) so those classes are known to the GraalVM AOT.

This will allow us to do many of the Java integration features we depend on, since the reflective parts will already have been handled.

* Build a new mechanism for doing dynamic calls that does not use invokedynamic, and instead uses the generated call sites above.

This eliminates the need for invokedynamic, which is unlikely to ever be supported on GraalVM.

**Comment by oleg-nenashev** [comment]
I might create a JRuby extension for Quarkus to resolve the Asciidoctor support issues in the JVM mode: https://github.com/asciidoctor/asciidoctorj/issues/1047

I will unlikely work on supporting GraalVM soon though. Reflection is still a very big obstacle though a Quarkus extension could be a good start there
